### PR TITLE
Selected shelf on "want to read" is deterministic

### LINKED ISSUE #829: In Firefox - can't remove items from My Reading Log after refresh
This issue is browser based, and happens in Firefox (does not reproduce in Chrome).

Reproduce:

1. Go to My Reading Log (make sure you have at least one item in the "Want to Read" list).
2. Items in the list will have "Want to Read + <check-mark>" in the drop down menu - expected behavior.
3. Refresh browser.
4. Items now show "Remove" (without the check-mark).
5. Reselecting "Remove" does not remove the item from the list.

image

### PR DESCRIPTION
**Pull request by jdlrobson**
Removing a book from a shelf should use a different option value from the other shelves - otherwise it's not clear to the browser which option value to render as selected.

Right now we're relying on the first result to be the selected one but this is what the selected property is for!

The template is DRYed up to be more readable and use the selected property rather than rely on element order.

The API is modified to accept and remove an entry from its current shelf when the value -1 is passed.

In the interest of simplifying the frontend, the number of books on a given shelf is no longer shown in the link label so we don't have to worry about updating the state of that as changes are made. In future we can revist this when we have a frontend framework that can track state.

Closes #829

### Testing
See the bug report for testing instructions.
You'll need to use Firefox.
Add a few books to your different reading lists to make sure all 3 want,  currently and already are set appropriately.

### CODE REVIEW

File: openlibrary/plugins/openlibrary/api.py [hunk 1]
```diff
@@ -151,9 +151,9 @@ def POST(self, work_id):
                 'error': 'Invalid bookshelf'
             }), content_type="application/json")
 
-        if bookshelf_id == current_status:
+        if bookshelf_id == current_status or bookshelf_id == -1:
             work_bookshelf = models.Bookshelves.remove(
```
Comment (mekarpeles) [inline openlibrary/plugins/openlibrary/api.py]: Have we tested this line in particular?

**Comment by jdlrobson** [comment]
Any updates on this ticket? I'm not sure what you need from me here.

File: openlibrary/macros/ReadingLogButton.html [hunk 2]
```diff
@@ -2,18 +2,18 @@
...
-    $elif read_status == 2:
-      <option value="2"><span class="activated-check">✓</span> Currently Reading</option>
-      <option value="1">Want to Read</option>
-      <option value="3">Already Read</option>
-    $else:
-      <option value="1"><span class="activated-check">✓</span> Want to Read</option>
-      <option value="2">Currently Reading</option>
-      <option value="3">Already Read</option>
-    <option value="$(read_status)">Remove</option>
+    $for value in [1,2,3]:
```
Comment (cdrini) [inline openlibrary/macros/ReadingLogButton.html]: Let's not reinvent a `dict` using `if/else` (again 😅 )

```suggestion
    $ shelves_by_id = {1: _('Want to Read'), 2: _('Currently Reading'), 3: _('Already Read'), -1: _('Remove') }
    $for shelf_id in [1, 2, 3, -1]:
        $ is_selected = read_status == shelf_id
        <option value="$shelf_id" $cond(is_selected, 'selected', '')>
        $ if is_selected:
            <span class="activated-check">✓</span>
        $shelves_by_id[shelf_id]
```

File: openlibrary/plugins/openlibrary/api.py [hunk 3]
```diff
@@ -151,9 +151,9 @@ def POST(self, work_id):
                 'error': 'Invalid bookshelf'
             }), content_type="application/json")
 
-        if bookshelf_id == current_status:
+        if bookshelf_id == current_status or bookshelf_id == -1:
```
Comment (cdrini) [inline openlibrary/plugins/openlibrary/api.py]: Don't need to support that anymore

```suggestion
        if bookshelf_id == -1:
```

File: openlibrary/templates/account/books.html [hunk 4]
```diff
@@ -64,9 +64,9 @@ <h3>
       <div class="mybooks-menu">
         <ul class="preset-lists">
           <li class="subsection">Reading Log</li>
-          <li><a href="$urlbase/currently-reading" $('class=selected' if key == 'currently-reading' else '')>Currently Reading ($(reading_log_counts['currently-reading']))</a></li>
-          <li><a href="$urlbase/want-to-read" $('class=selected' if key == 'want-to-read' else '')>Want to Read ($(reading_log_counts['want-to-read']))</a></li>
-          <li><a href="$urlbase/already-read" $('class=selected' if key == 'already-read' else '')>Already Read ($(reading_log_counts['already-read']))</a></li>
+          <li><a href="$urlbase/currently-reading" $('class=selected' if key == 'currently-reading' else '')>Currently Reading</a></li>
+          <li><a href="$urlbase/want-to-read" $('class=selected' if key == 'want-to-read' else '')>Want to Read</a></li>
+          <li><a href="$urlbase/already-read" $('class=selected' if key == 'already-read' else '')>Already Read</a></li>
```
Comment (cdrini) [inline openlibrary/templates/account/books.html]: Let's revert this change; as a user of this feature on GoodReads, I would rather have the numbers be off when I update the page than not have them there at all.

File: openlibrary/core/bookshelves.py [hunk 5]
```diff
@@ -6,6 +6,7 @@
 class Bookshelves(object):
 
     PRESET_BOOKSHELVES = {
+        'Remove from current list': -1,
```
Comment (cdrini) [inline openlibrary/core/bookshelves.py]: Since `-1` isn't a "real" shelf, we don't need this here; these values are only used for lower-level access (closer to DB). The `-1` only really applies at the UI/API level.

```suggestion
```

File: openlibrary/macros/ReadingLogButton.html [hunk 6]
```diff
@@ -2,18 +2,18 @@
 
 <form method="POST" action="$(work.key)/bookshelves.json?debug=true" class="reading-log-lite">
   <select name="bookshelf_id">
-    $if read_status == 3:
-      <option value="3"><span class="activated-check">✓</span> Already Read</option>
```
Comment (cdrini) [inline openlibrary/macros/ReadingLogButton.html]: Can you provide a screenshot of what this looks like without the check?

File: openlibrary/plugins/openlibrary/js/ol.js [hunk 7]
```diff
@@ -354,23 +354,22 @@ export function initReadingListFeature() {
     /* eslint-disable no-unused-vars */
     // success function receives data on successful request
     $(document).on('change', '.reading-log-lite select', function(e) {
-        var self = this;
-        var form = $(self).closest('form');
-        var remove = $(self).children('option').filter(':selected').text().toLowerCase() === 'remove';
-        var url = $(form).attr('action');
+        const self = this,
+            form = $(self).closest('form');
```
Comment (cdrini) [inline openlibrary/plugins/openlibrary/js/ol.js]: This is only used once; let's just inline it. It doesn't contribute to readability any.

```suggestion
```

File: openlibrary/plugins/openlibrary/js/ol.js [hunk 8]
```diff
@@ -354,23 +354,22 @@ export function initReadingListFeature() {
...
+
+        // On /account/books/want-to-read avoid a page reload by sending the
+        // new shelf to the server and removing the associated item.
+        // Note that any change to this select will result in the book changing
+        // shelf.
         $.ajax({
-            url: url,
+            url: $(form).attr('action'),
```
Comment (cdrini) [inline openlibrary/plugins/openlibrary/js/ol.js]:
```suggestion
            url: $self.closest('form').attr('action'),
```

File: openlibrary/plugins/openlibrary/js/ol.js [hunk 9]
```diff
@@ -354,23 +354,22 @@ export function initReadingListFeature() {
     /* eslint-disable no-unused-vars */
     // success function receives data on successful request
     $(document).on('change', '.reading-log-lite select', function(e) {
-        var self = this;
-        var form = $(self).closest('form');
-        var remove = $(self).children('option').filter(':selected').text().toLowerCase() === 'remove';
-        var url = $(form).attr('action');
+        const self = this,
```
Comment (cdrini) [inline openlibrary/plugins/openlibrary/js/ol.js]: This is only ever used as a jquery object

```suggestion
        const $self = $(this),
```

File: openlibrary/plugins/openlibrary/api.py (hunk 1 above)
Comment (cdrini) [reply openlibrary/plugins/openlibrary/api.py]: tested locally

File: openlibrary/macros/ReadingLogButton.html (hunk 2 above)
Comment (cdrini) [reply openlibrary/macros/ReadingLogButton.html]: Done

File: openlibrary/templates/account/books.html (hunk 4 above)
Comment (cdrini) [reply openlibrary/templates/account/books.html]: Done

File: openlibrary/core/bookshelves.py (hunk 5 above)
Comment (cdrini) [reply openlibrary/core/bookshelves.py]: Done

File: openlibrary/plugins/openlibrary/js/ol.js (hunk 7 above)
Comment (cdrini) [reply openlibrary/plugins/openlibrary/js/ol.js]: Done

File: openlibrary/plugins/openlibrary/js/ol.js (hunk 8 above)
Comment (cdrini) [reply openlibrary/plugins/openlibrary/js/ol.js]: Done

File: openlibrary/plugins/openlibrary/js/ol.js (hunk 9 above)
Comment (cdrini) [reply openlibrary/plugins/openlibrary/js/ol.js]: Done

**Comment by cdrini** [comment]
I'm going to try to fast-track this a bit to get it into today's deploy, if that's ok :)

File: openlibrary/macros/ReadingLogButton.html (hunk 6 above)
Comment (cdrini) [reply openlibrary/macros/ReadingLogButton.html]: Put it back

File: openlibrary/plugins/openlibrary/api.py (hunk 3 above)
Comment (cdrini) [reply openlibrary/plugins/openlibrary/api.py]: Unfortunately it looks like the `widget.html` file is still using this logic (searching for `/bookshelves.json` revealed). I think it also relies on `?action=remove` doing something (which it does not), so there's also that :/

Comment (cdrini) [reply openlibrary/plugins/openlibrary/api.py]: Won't fix

**Comment by cdrini** [comment]
@jdlrobson Ok! All the cr comments are addressed. Can I get a quick :+1: that this looks ok?

File: openlibrary/macros/ReadingLogButton.html [hunk 10]
```diff
@@ -2,18 +2,12 @@
...
-        $if read_status == value:
-          selected
-      >
-      $if value == 1:
-        Want to read
-      $elif value == 2:
-        Currently Reading
-      $elif value == 3:
-        Already Read
+    $ shelf_names = {1: _('Want to Read'), 2: _('Currently Reading'), 3: _('Already Read') }
```
Comment (jdlrobson) [inline openlibrary/macros/ReadingLogButton.html]: this looks much neater. thanks!

Comment (cdrini) [reply openlibrary/macros/ReadingLogButton.html]: 😊

**Comment by cdrini** [comment]
Tested on dev:
- ✅ Removing a book works
- ✅ Refreshing the page on firefox doesn't change everything to "Remove"
  - There is a _super_ edge case where if you remove a book, then hit refresh, the first item on the list will display "Remove" as selected. But don't think that's blocked.
